### DISCUSSION TITLE
RFC: Module API

### DISCUSSION THREAD

**Discussion by migueloller** [makeswift/makeswift#145] [Ideas]
# Background

Component registration in Makeswift currently involves doing something like this:

```tsx
import { ReactRuntime } from '@makeswift/runtime/react'
import { FancyComponent } from 'components/fancy-component'
import { Style } from '@makeswift/runtime/controls'

ReactRuntime.registerComponent(FancyComponent, {
  type: 'fancy-component',
  label: 'Fancy Component',
  props: {
    className: Style(),
  },
})

// And so on...
```

```tsx
import 'lib/makeswift'

// getStaticPaths, getStaticProps, etc.
```

There’s a couple of issues with this.

1. Integrating Makeswift requires creating a file just for registering components
2. The component registration file has to be imported in every page that uses Makeswift
3. Can’t do automatic optimizations, like code-splitting, because the registration API happens at runtime
4. Colocating component registration with components is very tedious and error prone

# Proposal

To fix these issues we propose a static API for registering components

```tsx
import { Style } from '@makeswift/runtime/controls'

export default function FancyComponent({ className }) {
  return <p className={className}>Fancy!</p>
}

export const makeswift = {
  type: 'fancy-component'
  label: 'Fancy Component'
  props: {
    className: Style()
  }
}
```

Instead of manually registering components with a runtime API (i.e., `ReactRuntime.registerComponent`) a module export, inspired by Next.js’ data fetching APIs (e.g., `getServerSideProps`), is used. This comes with the following benefits:

1. There’s no need to create a central file for registering components
2. Including the registered component’s code is done automatically for pages that use Makeswift
3. Code-splitting via `next/dynamic` can be done automatically
4. Makeswift integration code can be colocated directly in component files all while not impacting the component or the JS bundle

## Next.js Compiler Plugin & `makeswift.config.js`

To provide this new module API we will be implementing a plugin for the [Next.js Compiler](https://nextjs.org/docs/advanced-features/compiler). This plugin will identify modules that integrate with Makeswift and perform the necessary code transformations for the integration. The `makeswift.config.js` configuration file will allow full control of what modules are considered by the plugin.

For example, if you wanted to only consider files with the `makeswift.js` extension, then the `makeswift.config.js` file might look something like this:

```tsx
/** @type {import('@makeswift/next').MakeswiftConfig} */
module.exports = {
  components: ['./src/**/*.makeswift.{js,jsx,ts,tsx}'],
}
```

Or if you wanted to consider all files in the `src/components` folder, then maybe something like this:

```tsx
/** @type {import('@makeswift/next').MakeswiftConfig} */
module.exports = {
  components: ['./src/components/**/*.{js,jsx,ts,tsx}'],
}
```

Note that for both cases, if the modules don’t have a `makeswift` named export they won’t be considered. This could also be made configurable in case you wanted *all* modules to be included. For example, you might have a `src/components/makeswift` folder where every file is a Makeswift component.

The goal is to give full control and flexibility of where these modules can live and the ability to opt in with the `makeswift` named export.

# Backwards Compatibility

We will be deprecating the runtime API in favor of the new module API.

# Alternative Approaches

The only other alternative we’ve considered is the one that is currently implemented, the runtime API. We believe that the module API is strictly superior for all the reasons outlined above.

# Feedback

We’d love to gather as much feedback as possible on this RFC before we start implementation to make sure that the API covers all the relevant use cases.

**Comment by alexhwoods** [comment]
- How would extensions and components interact, if both are defined? The set intersection of the considered files?
- Imagine we had a type, FancyComponentProps. Could I get any value from having that type with my exported makeswift config? I could at least annotate whatever I export, right?

```typescript
type FancyComponentProps = { className: string }

// imagine we provide this type helper
type MakeswiftControls<T> = {
  [Prop in keyof T]: MakeswiftPropController;
};

const makeswiftProps: MakeswiftControls<FancyComponentProps> = {
  className: Style()
}

export const makeswift = {
  ...
  props: makeswiftProps
}
```

**Reply by migueloller** [reply]
> How would extensions and components interact, if both are defined? The set intersection of the considered files?

This is a good question. I think that the extensions should just be specified in the `components` option and that the `extensions` option should be removed. I'll update the RFC to reflect that.

> Imagine we had a type, FancyComponentProps. Could I get any value from having that type with my exported makeswift config? I could at least annotate whatever I export, right?
Perhaps we could do some sort of smart control registration based on types but this would likely happen from identifying the props of the default export, the component. If the component has types then it should be usable in the rest of the codebase without issue. I can't think of a specific benefit of exporting the types separately.

Having a type helper that infers the prop types from the prop schema sounds like a useful thing, though. We can definitely expose that in the public API.

Also, being able to share control definitions is very useful when composing components if you want to spread through props that could be provided by Makeswift.

**Reply by migueloller** [reply]
As for helping users with types, I was thinking we'd do something like this:

```ts
// components/button.tsx

type Props = {
  className: string
  children: string
}

export default function Button({ className, children }: Props) {
  return <button className={className}>{children}</button>
}

export const makeswift: MakeswiftComponentConfig<Props> = {
  type: 'button',
  label: 'Button',
  props: {
    className: Style(),
    children: Text({ defaultValue: 'Click me!' }),
  }
}
```

If a default value wasn't provided for the `children` prop in the `Text` control then there would be a type error indicating that the component expects a `string` for that field. So you would have to either provide a default value or make the prop be optional.

I think this is more inline with what you were suggesting. I misunderstood the question at first.

**Reply by alexhwoods** [reply]
Ok, sweet, yeah we're on the same page with that

**Comment by fikrikarim** [comment]
I like this API better, especially if this could fix the issues mentioned. 

I'm wondering if it's better to simplify the name, like `makeswift`, or make it more explicit, like `makeswiftRegisterComponent`.

**Reply by migueloller** [reply]
I personally like just `makeswift`. I think it's implicit that it's a component from the default export. Also, something like `makeswiftRegisterComponent` sounds imperative while this API is meant to be declarative

**Reply by fikrikarim** [reply]
Okay cool! I was thinking just in case we need to add similar API for other purposes.

**Reply by migueloller** [reply]
Any other Makeswift-related metadata or API would like be under the `makeswift` export I would think
